aptos-move/aptos-aggregator/src/resolver.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::aggregator_extension::AggregatorID;
+
+/// Returns a value of an aggregator from cache or global storage.
+///   - Ok(Some(..)) if the data exists
+///   - Ok(None)     if the data does not exist, only applies to Aggregator V1
+///                  which is a separate state item.
+///   - Err(..)      only when something really wrong happens.
+pub trait AggregatorResolver {
+    fn resolve_aggregator_value(&self, id: &AggregatorID) -> Result<Option<u128>, anyhow::Error>;
+}
+
+// Utils to store aggregator values in data store. Here, we
+// only care about aggregators which are state items.
+#[cfg(any(test, feature = "testing"))]
+pub mod test_utils {
+    use super::*;
+    use crate::{
+        aggregator_extension::AggregatorHandle,
+        delta_change_set::{deserialize, serialize},
+    };
+    use aptos_state_view::TStateView;
+    use aptos_types::state_store::{
+        state_key::StateKey, state_storage_usage::StateStorageUsage, state_value::StateValue,
+    };
+    use move_core_types::account_address::AccountAddress;
+    use move_table_extension::TableHandle;
+    use std::collections::HashMap;
+
+    /// Generates a dummy id for aggregator based on the given key. Only used for testing.
+    pub fn aggregator_id_for_test(key: u128) -> AggregatorID {
+        let bytes: Vec<u8> = [key.to_le_bytes(), key.to_le_bytes()]
+            .iter()
+            .flat_map(|b| b.to_vec())
+            .collect();
+        let key = AggregatorHandle(AccountAddress::from_bytes(bytes).unwrap());
+        AggregatorID::legacy(TableHandle(AccountAddress::ZERO), key)
+    }
+
+    #[derive(Default)]
+    pub struct AggregatorStore(HashMap<StateKey, StateValue>);
+
+    impl AggregatorStore {
+        pub fn set_from_id(&mut self, id: AggregatorID, value: u128) {
+            let state_key = id
+                .into_state_key()
+                .expect("Only table-based IDs are tested.");
+            self.set_from_state_key(state_key, value);
+        }
+
+        pub fn set_from_state_key(&mut self, state_key: StateKey, value: u128) {
+            self.0
+                .insert(state_key, StateValue::new_legacy(serialize(&value)));
+        }
+    }
+
+    impl AggregatorResolver for AggregatorStore {
+        fn resolve_aggregator_value(
+            &self,
+            id: &AggregatorID,
+        ) -> Result<Option<u128>, anyhow::Error> {
```

Comment:
Why does this output `Result<Option<u128>, anyhow::Error>` instead of `Result<u128, anyhow::Error>`?

Reply:
Aggregator can still be a table item which can be deleted (so storage returns None). Technically, we should not read something that has been deleted, only speculatively but this can just abort and will be re-executed anyway.  Let's use `Result<u128, anyhow::Error>`.